healthcheck.sh

```diff
@@ -0,0 +1,238 @@
+# --replication_name=n -> sets the multisource connection name tested
+# --replication_io -> IO thread is running
+# --replication_sql -> SQL thread is running
+# --replication_seconds_behind_master=n -> less than or equal this seconds of delay
+# --replication_sql_remaining_delay=n -> less than or equal this seconds of delay
+replication()
+{
+	# SHOW REPLICA available 10.5+
+	# https://github.com/koalaman/shellcheck/issues/2383
+	# shellcheck disable=SC2016,SC2026
+	_process_sql -e "show ${repl['all']:+all} slave${repl['name']:+s} ${repl['name']:+'${repl['name']}'} status\G" | \
+		{
+		shopt -s extglob
+		# Row header
+		read -t 5 -r
+		# read timeout
+		[ $? -gt 128 ] && return 1
+		while IFS=":" read -t 1 -r n v; do
+			n=${n##+([[:space:]])}
+			v=${v:1}
+			case "$n" in
+				Slave_IO_Running)
+					if [ -n "${repl['io']}" ] && [ "$v" = 'No' ]; then
+						return 1
+					fi
+					;;
+				Slave_SQL_Running)
+					if [ -n "${repl['sql']}" ] && [ "$v" = 'No' ]; then
+						return 1
+					fi
+					;;
+				Seconds_Behind_Master)
+					if [ -n "${repl['seconds_behind_master']}" ] &&
+						{ [ "$v" = NULL ] ||
+						[ "${repl['seconds_behind_master']}" -le "$v" ]; }; then
+						return 1
+					fi
+					;;
+				SQL_Remaining_Delay)
+					if [ -n "${repl['sql_remaining_delay']}" ] &&
+						{ [ "$v" = NULL ] ||
+						[ "${repl['sql_remaining_delay']}" -le "$v" ]; }; then
+						return 1
+					fi
+					;;
+			esac
+		done
+		# read timeout
+		[ $? -gt 128 ] && return 1
+		return 0
+       }
+       return $?
+}
+
+# mariadbupgrade
+#
+# Test the lock on the file /var/lib/mysql_upgrade_info
+# https://jira.mariadb.org/browse/MDEV-27068
+mariadbupgrade()
+{
+	local f='/var/lib/mysql/mysql_upgrade_info'
+	if [ -r "$f" ]; then
+		flock --exclusive --nonblock -n 9 9<$f
+		return $?
+	fi
+	return 0
+}
+
+
+# MAIN
+
+if [ $# -eq 0 ]; then
+	echo "At least one argument required" >/dev/stderr
+	exit 1
+fi
+
+declare -A repl
+
+_repl_param_check()
+{
+	case "$1" in
+		seconds_behind_master) ;&
+		sql_remaining_delay)
+			if [ -z "${repl['io']}" ]; then
+				repl['io']=1
+				echo "Forcing --replication_io=1, $1 requires IO thread to be running" > /dev/stderr
+			fi
+			;;
+		all)
+			if [ -z "${repl['name']}" ]; then
+				unset repl['name']
+				echo "Option --replication_all incompatible with specied source --replication_name, clearing replication_name" > /dev/stderr
+			fi
+			;;
+		name)
+			if [ -z "${repl['all']}" ]; then
+				unset repl['all']
+				echo "Option --replication_name incompatible with --replication_all, clearing replication_all" > /dev/stderr
+			fi
+			;;
+	esac
+}
+
+# Marks the end of mysql -> mariadb name changes in 10.6+
+#ENDOFSUBSTITUTIONS
+while [ $# -gt 0 ]; do
+	case "$1" in
+		--su=*)
+			u="${1#*-}"
+			shift
+			exec gosu "${u}" "${BASH_SOURCE[0]}" "$@"
+			;;
+		--su-mysql)
+			shift
+			exec gosu mysql "${BASH_SOURCE[0]}" "$@"
+			;;
+		--replication_*=*)
+			n=${1#*_}
+			n=${n%%=*}
+			n=${n,,*}
+			v=${1#*=}
```

Comment:
Personally I'd rather use `sed` here: It allows more stricter checking and is more readable. Same for `--replication_*` below.